File: ios/Stockbook/Features/Bills/BillTemplate.swift
import SwiftUI

/// One bill, drawn as the thing you turn round and show a customer.
///
/// The list rows and the cart show a bill in pieces — a summary line, a total, a
/// meta strip. This is the whole of it in one place: who wrote it, its number,
/// when, for whom, every line with its arithmetic visible, and what is still
/// owed. It is used both for the confirmation right after saving and for opening
/// any bill from history, so those two can never drift apart.
///
/// Every value here is the **snapshot taken at sale time**. A product renamed or
/// repriced since does not change what this says, which is the whole reason
/// `BillLine` carries its own name and price.
struct BillTemplate: View {
    let bill: Bill
    /// The shop's own name, from settings. Blank is fine — a shop that never
    /// entered one simply has no letterhead.
    var shopName: String = ""

    @Environment(\.currency) private var currency

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            heading

            FadedRule().padding(.vertical, 12)

            lines

            FadedRule().padding(.vertical, 12)

            totals
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(Nocturne.surface, in: RoundedRectangle(cornerRadius: Metrics.statRadius, style: .continuous))
        .hairline(radius: Metrics.statRadius)
    }

    // MARK: Letterhead

    private var heading: some View {
        VStack(alignment: .leading, spacing: 0) {
            if !shopName.isBlank {
                Kicker(shopName).padding(.bottom, 5)
            }

            HStack(alignment: .firstTextBaseline, spacing: 10) {
                Text(bill.reference(Loc))
                    .font(NocturneType.inter(20, .medium))
                Spacer(minLength: 8)
            }
            .padding(.bottom, 4)

            Text(Loc.billWhen(date: Loc.longDate(bill.createdAt), time: Loc.time(bill.createdAt)))
                .nocturneText(.meta)

            if !bill.who.isBlank {
                Text(Loc.billedTo(bill.who))
                    .nocturneText(.meta)
                    .padding(.top, 2)
            }
        }
    }

    // MARK: The items

    private var lines: some View {
        VStack(alignment: .leading, spacing: 10) {
            ForEach(bill.lines) { line in
                HStack(alignment: .firstTextBaseline, spacing: 10) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(line.name)
                            .font(NocturneType.inter(14))
                        // The arithmetic stays visible. A customer querying a
                        // total is nearly always querying one line's quantity
                        // or its price, and this is the answer without anyone
                        // having to recompute it at the counter.
                        Text(Loc.quantityAtPrice(
                            quantity: line.qty,
                            price: Money.text(line.price, in: currency)
                        ))
                        .nocturneText(.meta)
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)

                    Text(Money.text(line.lineTotal, in: currency))
                        .font(NocturneType.inter(14))
                }
            }
        }
    }

    // MARK: What it came to

    private var totals: some View {
        VStack(alignment: .leading, spacing: 0) {
            // What was knocked off, where anything was. The customer's own copy
            // of the bill is exactly where a discount belongs — it is the reason
            // the figure is what it is, and a shop that gave 10% away should get
            // the credit for it. The *statement* is the document that carries
            // only the total.
            if bill.isDiscounted {
                HStack {
                    Text(Loc.subtotalLabel).nocturneText(.meta)
                    Spacer()
                    Text(Money.text(bill.subtotal, in: currency))
                        .font(NocturneType.inter(12.5))
                        .foregroundStyle(Nocturne.neutral400)
                }
                HStack {
                    Text(Loc.discountOf(Money.amount(bill.discountPercent ?? 0, in: currency)))
                        .nocturneText(.meta)
                    Spacer()
                    Text("− " + Money.text(bill.discountAmount ?? 0, in: currency))
                        .font(NocturneType.inter(12.5))
                        .foregroundStyle(Nocturne.accent400)
                }
                .padding(.bottom, 6)
            }

            HStack(alignment: .firstTextBaseline) {
                Text(Loc.total)
                    .font(NocturneType.inter(13))
                    .foregroundStyle(Nocturne.neutral500)
                Spacer()
                Text(Money.text(bill.total, in: currency))
                    .font(NocturneType.inter(25, .medium))
                    .tracking(25 * -0.02)
            }

            Text(paymentNote)
                .font(NocturneType.inter(12.5))
                .foregroundStyle(Nocturne.accent400)
                .lineSpacing(3)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.top, 7)
        }
    }

    /// `Paid in full, cash.` or `Paid SAR 100 · Ahmed owes SAR 94`.
    private var paymentNote: String {
        guard let paid = bill.paid else { return Loc.paidInFullCash }
        return Loc.partPaidNote(
            paid: Money.text(paid, in: currency),
            who: bill.who,
            balance: Money.text(bill.balance, in: currency)
        )
    }
}

File: android/app/src/main/kotlin/com/stockbook/app/feature/bills/BillTemplate.kt
package com.stockbook.app.feature.bills

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.stockbook.app.design.FadedRule
import com.stockbook.app.design.Kicker
import com.stockbook.app.design.Metrics
import com.stockbook.app.design.Nocturne
import com.stockbook.app.design.NocturneType
import com.stockbook.app.design.card
import com.stockbook.app.design.hairline
import com.stockbook.core.model.Bill
import com.stockbook.core.model.Currency
import com.stockbook.core.money.Money
import com.stockbook.core.text.Strings

/**
 * One bill, drawn as the thing you turn round and show a customer.
 *
 * Used both for the confirmation right after saving and for opening any bill
 * from history, so those two can never drift apart.
 *
 * Every value here is the **snapshot taken at sale time**. A product renamed or
 * repriced since does not change what this says.
 */
@Composable
fun BillTemplate(
    bill: Bill,
    currency: Currency,
    strings: Strings,
    modifier: Modifier = Modifier,
    shopName: String = ""
) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .card(Metrics.statRadius)
            .hairline(radius = Metrics.statRadius)
            .padding(16.dp)
    ) {
        if (shopName.isNotBlank()) {
            Kicker(shopName, modifier = Modifier.padding(bottom = 5.dp))
        }

        // Nothing sits beside the reference any more. A bill entered wrongly is
        // corrected on the document itself, so there is no mark a bill can be
        // carrying by the time anybody reads one.
        Text(
            bill.reference(strings),
            style = NocturneType.inter(20.0, FontWeight.Medium),
            color = Nocturne.text,
            modifier = Modifier.fillMaxWidth()
        )

        Text(
            strings.billWhen(strings.longDate(bill.createdAt), strings.time(bill.createdAt)),
            style = NocturneType.meta,
            color = Nocturne.neutral500,
            modifier = Modifier.padding(top = 4.dp)
        )
        if (bill.who.isNotBlank()) {
            Text(
                strings.billedTo(bill.who),
                style = NocturneType.meta,
                color = Nocturne.neutral500,
                modifier = Modifier.padding(top = 2.dp)
            )
        }

        // One rule where there is nothing between them, two where there is. A bill
        // entered as a figure has no lines to list, and the empty band of nothing
        // that a second rule opened up read as a document that had lost its
        // contents rather than one that never had any.
        FadedRule(modifier = Modifier.padding(vertical = 12.dp))

        if (bill.isItemised) {
            bill.lines.forEachIndexed { index, line ->
                if (index > 0) Spacer(Modifier.height(10.dp))
                Row(modifier = Modifier.fillMaxWidth()) {
                    Column(modifier = Modifier.weight(1f)) {
                        Text(line.name, style = NocturneType.inter(14.0), color = Nocturne.text)
                        // The arithmetic stays visible. A customer querying a total
                        // is nearly always querying one line's quantity or price,
                        // and this is the answer without anyone recomputing it.
                        Text(
                            strings.quantityAtPrice(line.qty, Money.text(line.price, currency)),
                            style = NocturneType.meta,
                            color = Nocturne.neutral500,
                            modifier = Modifier.padding(top = 2.dp)
                        )
                    }
                    Spacer(Modifier.width(10.dp))
                    Text(
                        Money.text(line.lineTotal, currency),
                        style = NocturneType.inter(14.0),
                        color = Nocturne.text
                    )
                }
            }

            FadedRule(modifier = Modifier.padding(vertical = 12.dp))
        }

        Row(verticalAlignment = Alignment.Bottom, modifier = Modifier.fillMaxWidth()) {
            Text(
                strings.total,
                style = NocturneType.inter(13.0),
                color = Nocturne.neutral500,
                modifier = Modifier.weight(1f)
            )
            Text(
                Money.text(bill.total, currency),
                style = NocturneType.bigNumber(25.0),
                color = Nocturne.text
            )
        }

        Text(
            paymentNote(bill, currency, strings),
            style = NocturneType.inter(12.5),
            color = Nocturne.accent400,
            modifier = Modifier.padding(top = 7.dp)
        )
    }
}

/**
 * Settled at the counter, or what is left and who owes it.
 */
private fun paymentNote(bill: Bill, currency: Currency, strings: Strings): String {
    val paid = bill.paid ?: return strings.paidInFullCash
    return strings.partPaidNote(
        paid = Money.text(paid, currency),
        who = bill.who,
        balance = Money.text(bill.balance, currency)
    )
}
